### DISCUSSION TITLE
Query params with encoded ampersand (%26) parsed incorrectly?

### DISCUSSION THREAD

**Discussion by matanper** [fastapi/fastapi#11793] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def root(param: str | None = None):
    return {"param": param}
```

### Description

Calling the endpoint with `curl -X GET "http://localhost:8000?param=foo%26other=bar"` 
results in `{"param":"foo&other=bar"}` instead of `{"param":"foo"}`
Sometimes I get encoded requests like that to my API and sadly I can't control it. Is there any easy fix? 

### Operating System

macOS

### FastAPI Version

0.111.0

### Pydantic Version

2.7.4

### Python Version

3.10.6

**Comment by matanper** [comment]
This middleware I wrote seems to work, but I'm unsure if always doing unqoute can corrupt the query string
```
class QueryParamsMiddleware:
    def __init__(self, app: ASGIApp):
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send):
        if scope['type'] == 'http':
            request = Request(scope)
            qs = request.scope['query_string']
            request.scope['query_string'] = urllib.parse.unquote(qs).encode()
        return await self.app(scope, receive, send)
```

**Reply by ceb10n** [reply]
Although it may be a problem with a fastapi app, it is how starlette deals with query params.

Starlette's `request_response` will pass a `Request` to fastapi's `APIRoute`.

A `Request` is an `HTTPConnection` that has a `query_params` property:

```python
    @property
    def query_params(self) -> QueryParams:
        if not hasattr(self, "_query_params"):
            self._query_params = QueryParams(self.scope["query_string"])
        return self._query_params
```

It will create `QueryParams` relying on `urllib.parse.parse_qsl`, that will split your query string with `&`.

My opinion: If it is really a problem, maybe it should be fixed in starlette.

If Starlette's team don't think it's a problem, then maybe you should create your own middleware to deal with encoded ampersand like @matanper showed us.

**Comment by YuriiMotov** [comment]
This is definitely not something that should be fixed in neither FastAPI nor Starlette.

You can use the solution like suggested above on your own risk. But keep in mind that in some cases it can brake your requests. Imagine you need to send `&` character as a value of `param` query field. In normal way you will urlencode it. But your middleware will decode it and you will have something like `http://localhost:8000?param=&&other=bar`. This way `param` will be empty.
